### PR TITLE
Save build artifacts for all configurations (among other things)

### LINKED ISSUE #203: Build script only works for one configuration
**Describe the bug**
Currently, the script to build MFC is a one way street. For instance, if call the script after you've built it once like this:

```
% ./mfc.sh build --gpu
mfc: OK > (venv) Entered the Python virtual environment.
Switching from mpi=Yes & gpu=No & debug=No to mpi=Yes & gpu=Yes & debug=No

Removing build/pre_process
Removing build/simulation
Removing build/post_process
Removing build/syscheck
                                      
      .=++*:          -+*+=.          magbad@MagBad-WS [Linux]
     :+   -*-        ==   =* .        ------------------------
   :*+      ==      ++    .+-         --jobs 1
  :*##-.....:*+   .#%+++=--+=:::.     --mpi
  -=-++-======#=--**+++==+*++=::-:.   --gpu
 .:++=----------====+*= ==..:%.....   --no-debug
  .:-=++++===--==+=-+=   +.  :=       --targets pre_process, simulation, and post_process
  +#=::::::::=%=. -+:    =+   *:      -----------------------------------------------------------
 .*=-=*=..    :=+*+:      -...--      $ ./mfc.sh [build, run, test, clean, count, packer] --help

```

You'd remove all the previous build process. Though the building, MFC doesn't take that long (about 5 minutes), it deletes all the downloaded & build dependencies.

In addition, the build process may abort in case the passed parameters are wrong. Removing the `--gpu` flag from the above call result in:

```
./mfc.sh build 
mfc: OK > (venv) Entered the Python virtual environment.
                                      
      .=++*:          -+*+=.          magbad@MagBad-WS [Linux]
     :+   -*-        ==   =* .        ------------------------
   :*+      ==      ++    .+-         --jobs 1
  :*##-.....:*+   .#%+++=--+=:::.     --mpi
  -=-++-======#=--**+++==+*++=::-:.   --gpu
 .:++=----------====+*= ==..:%.....   --no-debug
  .:-=++++===--==+=-+=   +.  :=       --targets pre_process, simulation, and post_process
  +#=::::::::=%=. -+:    =+   *:      -----------------------------------------------------------
 .*=-=*=..    :=+*+:      -...--      $ ./mfc.sh [build, run, test, clean, count, packer] --help

Building pre_process:

  $ cmake -DMFC_PRE_PROCESS=ON -Wno-dev -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_BUILD_TYPE=Release 
-DCMAKE_PREFIX_PATH=/home/magbad/opt/MFC/build/install -DCMAKE_FIND_ROOT_PATH=/home/magbad/opt/MFC/build/install 
-DCMAKE_INSTALL_PREFIX=/home/magbad/opt/MFC/build/install -DMFC_MPI=ON -DMFC_OpenACC=ON -S /home/magbad/opt/MFC/ -B 
/home/magbad/opt/MFC/build/pre_process

-- The C compiler identification is GNU 11.4.0
-- The Fortran compiler identification is GNU 11.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - done
-- Check for working Fortran compiler: /usr/bin/gfortran - skipped
-- Performing Test SUPPORTS_MARCH_NATIVE
-- Performing Test SUPPORTS_MARCH_NATIVE - Success
-- Enabled IPO / LTO
-- Found MPI_Fortran: /usr/lib/x86_64-linux-gnu/libmpi_usempif08.so (found version "3.1") 
-- Found MPI: TRUE (found version "3.1") found components: Fortran 
-- Configuring done
-- Generating done
-- Build files have been written to: /home/magbad/opt/MFC/build/pre_process
  
  Generating pre_process/include/case.fpp.
    INFO: Existing case.fpp file is up to date.

  $ cmake --build /home/magbad/opt/MFC/build/pre_process --target pre_process -j 1 --config Release

Scanning dependencies of target pre_process
[  3%] Building Fortran object CMakeFiles/pre_process.dir/src/pre_process/autogen/m_constants.fpp.f90.o
[  6%] Building Fortran object CMakeFiles/pre_process.dir/src/pre_process/autogen/m_derived_types.fpp.f90.o
[  9%] Building Fortran object CMakeFiles/pre_process.dir/src/pre_process/autogen/m_global_parameters.fpp.f90.o
[ 12%] Building Fortran object CMakeFiles/pre_process.dir/src/pre_process/autogen/m_mpi_common.fpp.f90.o
[ 15%] Building Fortran object CMakeFiles/pre_process.dir/src/pre_process/autogen/m_mpi_proxy.fpp.f90.o
[ 18%] Building Fortran object CMakeFiles/pre_process.dir/src/common/m_compile_specific.f90.o
[ 21%] Building Fortran object CMakeFiles/pre_process.dir/src/common/m_eigen_solver.f90.o
[ 24%] Building Fortran object CMakeFiles/pre_process.dir/src/pre_process/autogen/m_helper.fpp.f90.o
[ 27%] Building Fortran object CMakeFiles/pre_process.dir/src/pre_process/autogen/m_variables_conversion.fpp.f90.o
[ 30%] Building Fortran object CMakeFiles/pre_process.dir/src/pre_process/autogen/m_data_output.fpp.f90.o
[ 33%] Building Fortran object CMakeFiles/pre_process.dir/src/pre_process/autogen/m_check_patches.fpp.f90.o
[ 36%] Building Fortran object CMakeFiles/pre_process.dir/src/pre_process/m_assign_variables.f90.o
[ 39%] Building Fortran object CMakeFiles/pre_process.dir/src/pre_process/autogen/m_model.fpp.f90.o
[ 42%] Building Fortran object CMakeFiles/pre_process.dir/src/pre_process/autogen/m_patches.fpp.f90.o
[ 45%] Building Fortran object CMakeFiles/pre_process.dir/src/pre_process/autogen/m_initial_condition.fpp.f90.o
[ 48%] Building Fortran object CMakeFiles/pre_process.dir/src/pre_process/m_checker.f90.o
[ 51%] Building Fortran object CMakeFiles/pre_process.dir/src/pre_process/m_grid.f90.o
[ 54%] Building Fortran object CMakeFiles/pre_process.dir/src/pre_process/autogen/m_start_up.fpp.f90.o
[ 57%] Building Fortran object CMakeFiles/pre_process.dir/src/pre_process/p_main.f90.o
[ 60%] Linking Fortran executable pre_process
lto-wrapper: warning: using serial compilation of 4 LTRANS jobs
[100%] Built target pre_process

  $ cmake --install /home/magbad/opt/MFC/build/pre_process

-- Install configuration: "Release"
-- Installing: /home/magbad/opt/MFC/build/install/bin/pre_process

Building syscheck:

  $ cmake -DMFC_SYSCHECK=ON -Wno-dev -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_BUILD_TYPE=Release 
-DCMAKE_PREFIX_PATH=/home/magbad/opt/MFC/build/install -DCMAKE_FIND_ROOT_PATH=/home/magbad/opt/MFC/build/install 
-DCMAKE_INSTALL_PREFIX=/home/magbad/opt/MFC/build/install -DMFC_MPI=ON -DMFC_OpenACC=ON -S /home/magbad/opt/MFC/ -B 
/home/magbad/opt/MFC/build/syscheck

-- The C compiler identification is GNU 11.4.0
-- The Fortran compiler identification is GNU 11.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - done
-- Check for working Fortran compiler: /usr/bin/gfortran - skipped
-- Performing Test SUPPORTS_MARCH_NATIVE
-- Performing Test SUPPORTS_MARCH_NATIVE - Success
-- Enabled IPO / LTO
-- Found MPI_Fortran: /usr/lib/x86_64-linux-gnu/libmpi_usempif08.so (found version "3.1") 
-- Found MPI: TRUE (found version "3.1") found components: Fortran 
-- Found OpenACC_C: -fopenacc  
-- Found OpenACC_Fortran: -fopenacc  
-- Configuring done
-- Generating done
-- Build files have been written to: /home/magbad/opt/MFC/build/syscheck
  
  No additional input file generation needed for syscheck.

  $ cmake --build /home/magbad/opt/MFC/build/syscheck --target syscheck -j 1 --config Release

Scanning dependencies of target syscheck
[ 33%] Building Fortran object CMakeFiles/syscheck.dir/src/syscheck/autogen/syscheck.fpp.f90.o
gfortran: error: unrecognized command-line option ‘-foffload-options=-lgfortran -lm’
gmake[3]: *** [CMakeFiles/syscheck.dir/build.make:79: CMakeFiles/syscheck.dir/src/syscheck/autogen/syscheck.fpp.f90.o] Error 1
gmake[2]: *** [CMakeFiles/Makefile2:83: CMakeFiles/syscheck.dir/all] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:90: CMakeFiles/syscheck.dir/rule] Error 2
gmake: *** [Makefile:170: syscheck] Error 2


Error: Failed to build the syscheck target.

Terminated
```

**To Reproduce**
See description for the example

**Expected behavior**
IMHO, a tree like this:

```
build/releaseMpi<other-flags>
build/debug<...>
```

+ a symlink to the latest build in the project root would be a good solution. 
This structure would enable developers to reuse different environments, therefore increasing their productivity.

**Environment**
NA

**Additional context**

### PR DESCRIPTION
**Pull request by henryleberre**
Closes #203, #206, #209, #210.

### CODE REVIEW

File: toolchain/mfc/build.py [hunk 1]
```diff
@@ -66,23 +65,43 @@ def get_target(name: str) -> MFCTarget:
...
 
 # Get the directory that contains the target's CMakeLists.txt
 def get_cmake_dirpath(target: MFCTarget) -> str:
-    subdir = ["", os.sep.join(["toolchain", "dependencies"])][int(target.isDependency)]
+    return os.sep.join([
+        os.getcwd(),
+        ["", os.sep.join(["toolchain", "dependencies"])][int(target.isDependency)]
+    ])
 
-    return os.sep.join([os.getcwd(), subdir])
```
Comment (sbryngelson) [inline toolchain/mfc/build.py:86]: comment goes here

File: toolchain/mfc/build.py [hunk 2]
```diff
@@ -66,23 +65,43 @@ def get_target(name: str) -> MFCTarget:
...
+    return os.sep.join([
+        os.getcwd(),
+        "build",
+        "install",
+        [CFG().make_slug(), 'dependencies'][int(target.isDependency)]
+    ])
 
-def get_install_dirpath() -> str:
-    return os.sep.join([os.getcwd(), "build", "install"])
+
```
Comment (sbryngelson) [inline toolchain/mfc/build.py:98]: comment goes here

File: toolchain/mfc/build.py [hunk 3]
```diff
@@ -66,23 +65,43 @@ def get_target(name: str) -> MFCTarget:
...
+def get_dependency_install_dirpath() -> str:
+    for target in TARGETS:
+        if target.isDependency:
+            return get_install_dirpath(target)
+
+    raise common.MFCException("No dependency target found.")
 
 
 def is_target_configured(target: MFCTarget) -> bool:
```
Comment (sbryngelson) [inline toolchain/mfc/build.py:109]: comment goes here

File: toolchain/mfc/run/engines.py [hunk 4]
```diff
@@ -44,7 +44,7 @@ def run(self, names: typing.List[str]) -> None:
         raise common.MFCException(f"MFCEngine::run: not implemented for {self.name}.")
 
     def get_binpath(self, target: str) -> str:
```
Comment (sbryngelson) [inline toolchain/mfc/run/engines.py:46]: comment goes here

File: toolchain/mfc/run/engines.py [hunk 5]
```diff
@@ -44,7 +44,7 @@ def run(self, names: typing.List[str]) -> None:
...
-        return os.sep.join([build.get_install_dirpath(), "bin", target])
+        return os.sep.join([build.get_install_dirpath(build.get_target(target)), "bin", target])
 
 
 def _interactive_working_worker(cmd, q):
```
Comment (sbryngelson) [inline toolchain/mfc/run/engines.py]: comment goes here

File: toolchain/mfc/state.py [hunk 6]
```diff
@@ -15,6 +15,15 @@ def from_dict(d: dict):
 
         return r
```
Comment (sbryngelson) [inline toolchain/mfc/state.py:19]: comment goes here

File: toolchain/mfc/state.py [hunk 7]
```diff
@@ -15,6 +15,15 @@ def from_dict(d: dict):
...
 
+    def items(self) -> typing.List[typing.Tuple[str, bool]]:
+        return { field.name: getattr(self, field.name) for field in dataclasses.fields(self) }.items()
+
```
Comment (sbryngelson) [inline toolchain/mfc/state.py:22]: comment goes here

File: toolchain/mfc/state.py [hunk 8]
```diff
@@ -15,6 +15,15 @@ def from_dict(d: dict):
...
+    def make_options(self) -> typing.List[str]:
+        return [ f"--{'no-' if not v else ''}{k}" for k, v in self.items() ]
+
```
Comment (sbryngelson) [inline toolchain/mfc/state.py:27]: comment goes here

File: toolchain/mfc/build.py [hunk 9]
```diff
@@ -28,17 +28,16 @@ def compute(self) -> typing.List[str]:
     requires:     Dependencies
```
Comment (sbryngelson) [inline toolchain/mfc/build.py:30]: comment goes here (what do the arguments mean?)

**Review by sbryngelson — CHANGES_REQUESTED** [review]

**Comment by sbryngelson** [comment]
alternative to adding a bunch of comments is to rewrite all of the one-liners into more expressive multi-line logic.

**Comment by sbryngelson** [comment]
This passes on GPU now. Do we need new goldenfiles for `weno_avg`? Presumably, they were originally generated via CPU by @haochey , so changing GPU behavior shouldn't change the 'correct answer' or goldenfile.

**Comment by haochey** [comment]
> This passes on GPU now. Do we need new goldenfiles for `weno_avg`? Presumably, they were originally generated via CPU by @haochey , so changing GPU behavior shouldn't change the 'correct answer' or goldenfile.

Hi @sbryngelson , Yes, the golden files with weno_avg were originally genearted via CPU with GNU and test on GPU on Phoenix passes too.

**Comment by sbryngelson** [comment]
@haochey they started failing (see here https://github.com/MFlowCode/MFC/actions/runs/5969092038/job/16194270272?pr=204) once @henryleberre fixed part of the test suite. Then @henryleberre fixed the GPU code (see here https://github.com/MFlowCode/MFC/pull/204/commits/ec688966757a5fdd882a794b46aef348878a4c18). Thank you @henryleberre !

**Comment by haochey** [comment]
> @haochey they started failing (see here https://github.com/MFlowCode/MFC/actions/runs/5969092038/job/16194270272?pr=204) once @henryleberre fixed part of the test suite. Then @henryleberre fixed the GPU code (see here [ec68896](https://github.com/MFlowCode/MFC/commit/ec688966757a5fdd882a794b46aef348878a4c18)). Thank you @henryleberre !

Thanks! @henryleberre

**Pull request merged**